astroquery/xmm_newton/core.py

```diff
@@ -0,0 +1,343 @@
+        filter : string
+            filter, optional
+            examples: Closed, Open, Thick, UVM2, UVW1, UVW2, V
+        expflag : string
+            exposure flag, optional, by default everything
+            values: S, U, X(not applicable)
+        expno : integer
+            exposure number with 3 digits, by default all exposures
+            examples: 001, 003
+        name : string
+            product type, optional, 6 characters, by default all product types
+            examples: 3COLIM, ATTTSR, EVENLI, SBSPEC, EXPMAP, SRCARF
+        datasubsetno : character
+            data subset number, optional, by default all
+            examples: 0, 1
+        sourceno : hex value
+            source number, optional, by default all sources
+            example: 00A, 021, 001
+        extension : string
+            file format, optional, by default all formats
+            values: ASC, ASZ, FTZ, HTM, IND, PDF, PNG
+        filename : string
+            file name to be used to store the file, optional, default
+            None
+        verbose : bool
+            optional, default 'False'
+            flag to display information about the process
+
+        Returns
+        -------
+        None. It downloads the observation indicated
+        """
+
+        link = self.data_aio_url + "obsno=" + observation_id
+
+        if level is not None:
+            link = link + "&level=" + level
+
+        if instname is not None:
+            link = link + "&instname=" + instname
+
+        if instmode is not None:
+            link = link + "&instmode=" + instmode
+
+        if filter is not None:
+            link = link + "&filter=" + filter
+
+        if expflag is not None:
+            link = link + "&expflag=" + expflag
+
+        if expno is not None:
+            link = link + "&expno=" + expno
+
+        if name is not None:
+            link = link + "&name=" + name
+
+        if datasubsetno is not None:
+            link = link + "&datasubsetno=" + datasubsetno
+
+        if sourceno is not None:
+            link = link + "&sourceno=" + sourceno
+
+        if extension is not None:
+            link = link + "&extension=" + extension
+
+        response = self._handler.request('GET', link)
+        if response is not None:
+            response.raise_for_status()
+
+            if filename is None:
+                if "Content-Disposition" in response.headers.keys():
+                    filename = re.findall('filename="(.+)"',
+                                          response.headers[
+                                              "Content-Disposition"])[0]
+                else:
+                    filename = observation_id + ".tar"
+
+            self._handler.get_file(filename, response=response,
+                                   verbose=verbose)
+
+            if verbose:
+                log.info("Wrote {0} to {1}".format(link, filename))
+
+            return filename
```

Comment:
Instead of a `download_data` function, this should be a `get_data` (`get_images`? `get_catalog`? whatever) function and it should return the downloaded file object in memory.  The user should be given the option whether they want to write the data to disk; this function does not give them that option.

Reply:
This has historical reason in XSA. After talking with the XMM people they prefer refer to download when the fits, pipeline products or postcards are downloaded and use the get for metadata, than can be retrieved in different formats (VOTable, CSV..). Initially I have download_product and they ask me to change to download_data as it refers to ODF files and also to pipeline products.